### PR TITLE
Implement instance API methods

### LINKED ISSUE #294: Implement Instance API
Mastodon's Instance API allows a user to discover information about a Mastodon website. The goal of this task is to implement the remaining endpoints of this API in BigBone:

- List of connected domains
- Weekly activity
- List of rules
- View moderated servers
- View extended description

The Instance API as it is documented by Mastodon can be found here:
https://docs.joinmastodon.org/methods/instance/

Tip: A lot of other endpoints similar to this have already been implemented. In order to get an idea how to implement this, I suggest that you study the code upfront. Start with one of the `*Methods.kt` files.

Please consult the following documents to make sure your contribution meets our projects criteria:

- [Code of Conduct](https://github.com/andregasser/bigbone/blob/master/CODE_OF_CONDUCT.md)
- [Code Contribution Guidelines](https://github.com/andregasser/bigbone/blob/master/CONTRIBUTING.md)

### PR DESCRIPTION
**Pull request by PattaFeuFeu**
# Description

This PR adds previously missing API methods under the [instance API methods](https://docs.joinmastodon.org/methods/instance/):

* View server information (V2)
* List of connected domains
* Weekly activity
* List of rules
* View moderated servers
* View extended description

Closes #294

## Breaking changes

There’s a major breaking change here:

I have removed the deprecated “View server information(V1)” method and replaced it with the new “View server information” (V2) variant, but under the same method name.

The `Instance` entity underwent major changes that are incompatible with previous calls.

# Type of Change

- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- This change requires a documentation update
  - [API Coverage](https://github.com/andregasser/bigbone/wiki/Mastodon-API-Coverage) page needs to be updated 

# How Has This Been Tested?

* New unit tests
* `gradle check`

# Mandatory Checklist

- [x] My change follows the projects coding style
- [x] I ran `gradle check` and there were no errors reported
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

# Optional Things To Check

- [x] In case you worked on new features: Did you also do the Rx part?
- [x] In case you worked on endpoints: Please mention in the PR that [API Coverage](https://github.com/andregasser/bigbone/wiki/Mastodon-API-Coverage) page needs to be updated (if needed)

### CODE REVIEW

**Comment by bocops** [comment]
> I have removed the deprecated “View server information(V1)” method and replaced it with the new “View server information” (V2) variant, but under the same method name.

The current idea is fro this library to support two major versions of Mastodon server (3.x and 4.x), and the V2 variant here was only added with 4.0.0. If we actually want to keep doing this, we might need an implementation of both endpoints and Instance entities.

That being said, we actually haven't got around to doing this anywhere else, plus Mastodon 4.x is [stated to be approaching its EOL](https://mastodon.social/@MastodonEngineering/111211854006817590), so I'd suggest to relax this requirement a bit for the time being.

This might need some more discussion, though. @andregasser what do you think?

**Comment by PattaFeuFeu** [comment]
@bocops I was wondering the same and had written a private message on our Slack to @andregasser about that. Until we have a released version, I’d be okay with more or less breaking changes, but starting with our first final release with a version number, we should introduce some concept of deprecation.

It could also be a good idea to add version info to individual methods and entities (or their properties) so that you can see on which version something should be available and on which not—on a server running v3, a v4+ attribute would be null even though it’s non-optional on v4+, so should that mean that attribute should still be nullable?

**Comment by bocops** [comment]
That's definitely reasonable. I'll comment on Slack regarding our version support strategy. In the meantime, I have no objections against going forward with this, but we should probably edit https://github.com/andregasser/bigbone/wiki/API-Support accordingly.

**Comment by PattaFeuFeu** [comment]
@bocops I could also re-add the old method and create an InstanceV1 entity and slap a `@Deprecated` annotation on both of them, if necessary.

But yeah, we should come up with some requirements. The document you linked seems more like wishful thinking than would we currently have in the repo. 😄 I do agree with those ideas though.

**Comment by andregasser** [comment]
Good morning guys, I need some time to review this (and the comments associated with it on Slack). I'll get back to you soon.

**Comment by andregasser** [comment]
I suggest that we take the time to revise the current API Support document. Most of it is not implemented and is just a collection of thoughts on how it **could** be handled. I will take a look at this in the coming days.

Regarding this PR:
I would keep the implementation from @PattaFeuFeu for now, but at the same time provide the `getInstanceV1` method. It should not take too much effort to do this and at least we cover the 3.x Mastodon servers (if there's still some  we don't know for sure).

Also please marke the `getInstanceV1` method as deprecated.

After that, we can merge this into master.

File: bigbone/src/main/kotlin/social/bigbone/api/entity/Instance.kt
Comment (andregasser) [inline bigbone/src/main/kotlin/social/bigbone/api/entity/Instance.kt:459]: Thinking about these ISO8601 formatted date strings we send and receive; Wouldn't it be nice, if we could directly convert this into an `Instant` instead? This would relieve BigBone users from having to do it and make using our lib more convenient. 

If we plan to to this, this should happen as part of a separate PR and the whole set of entities.

**Review by andregasser — CHANGES_REQUESTED** [review]
First off, thanks for your contribution here. I've added a couple of comments, that you should have a look at. But overall it is looking good!

File: bigbone/src/main/kotlin/social/bigbone/api/entity/Instance.kt
Comment (PattaFeuFeu) [reply bigbone/src/main/kotlin/social/bigbone/api/entity/Instance.kt:459]: Yep, I was thinking the same when I saw these date strings.

I’ll create a ticket and get on with implementing that next! 😊

**Comment by PattaFeuFeu** [comment]
@andregasser Done! 😊

**Review by andregasser — APPROVED** [review]
Looks good!

**Pull request merged**
